Lushu/src/main/kotlin/lushu/ContextGrammar/Grammar/Rules.kt

```diff
@@ -0,0 +1,226 @@
+            return (inputTokens.size - index - 1)
+        }
+        if (current.isPlus()) {
+            val (nextIndex: Int, nextNode: Node?) = plusCaseMatcher(inputTokens, mutableTokens, index, current)
+
+            if ((nextNode != null) && (nextNode.isSensitive()))
+                mutableTokens[nextIndex] = "*".repeat(mutableTokens[nextIndex].length)
+
+            return matchTokensAgainstPatternContext(inputTokens, mutableTokens, nextIndex, nextNode)
+        }
+        else {
+            val nextNode = matcher(current, inputTokens[index])
+
+            if ((nextNode != null) && (nextNode.isSensitive()))
+                mutableTokens[index] = "*".repeat(mutableTokens[index].length)
+
+            return matchTokensAgainstPatternContext(inputTokens, mutableTokens, index + 1, nextNode)
+        }
+    }
+
+    /**
+     * Finds the matching indices of tokens in the input list based on a list of regular expressions.
+     * @param inputTokens The list of tokens to search for matches.
+     * @param regexList The list of regular expressions to match against the tokens.
+     * @return A list of unique and sorted indices of the matching tokens.
+     * Example:
+     * Input: ["error", "warning", "exception", "success", "invalid"], ["e.", "s.", "c."]
+     * Output: [0, 2, 3]
+     */
+    private fun findMatchingIndex(inputTokens: List<String>, regexList: List<String>): List<Int> {
+
+        val matchingIndex = regexList.flatMap { regex ->
+            val pattern = Pattern.compile(regex)
+            inputTokens.mapIndexedNotNull { index, word ->
+                val matcher = pattern.matcher(word)
+                if (matcher.find()) index else null
+            }
+        }
+        return matchingIndex.distinct().sorted()
+    }
+
+    /**
+     * Converts a list of tokens into cipher tokens.
+     * This function iterates over the input list and checks if it contains any text patterns that need to be encrypted.
+     * The resulting cipher tokens are returned as a list.
+     * @param inputTokens The list of input tokens to be converted into cipher tokens.
+     * @return The list of cipher tokens generated from the input tokens.
+     * Example:
+     * Let a context "host ip: 123.123.123.123" where 123.123.123.123 needs to be encrypted.
+     * Input: ["server", "host", "ip:",  "123.123.123.123", "is", "online"]
+     * Output: ["server", "host", "ip:",  "***************", "is", "online"]
+     */
+    fun tokens2CipherTokens(inputTokens: List<String>): List<String>{
+        val regexList: List<String> = root.getChildren().map { child -> child.getRegex() }
+        val matchingIndex: List<Int> = findMatchingIndex(inputTokens, regexList)
+
+        var i = -1
+
+        val cipherTokens = inputTokens.toMutableList()
+
+        matchingIndex.forEach { index ->
+            if(i <= index){
+                i = matchTokensAgainstPatternContext(cipherTokens, inputTokens.toMutableList(), index, root)}
```

Comment:
```suggestion
            if(i <= index) {
                i = matchTokensAgainstPatternContext(cipherTokens, inputTokens.toMutableList(), index, root)
            }
```
Please use `ktlint` to deal with these issues automatically.